### Stage C Quiz Answers

##### Kindly note that my browser did not scroll at all during the period of the quiz. Though it has happened before, I was able to use the "Tab" key on my keyboard to somehow navigate back then. Unfortunately, even the Tab key did not work this time around. As such, I was unable to read most questions properly (fully) or even attempt them, especially he non-coding questions. Kindly help look into what could be the cause of the issue. Thank you.

#### Preliminary Analysis

In [65]:
#Import Pandas Library
import pandas as pd

In [66]:
#Read the CSV file containing dataset
df = pd.read_csv('C:/Users/hp/Downloads/Hamoye/Stage C _ ML Classfcatn/Stage C Tag-Along Project/Data_for_UCI_named.csv')

In [67]:
#View the table
df

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
...        ...       ...       ...       ...       ...       ...       ...   
9995  2.930406  9.487627  2.376523  6.187797  3.343416 -0.658054 -1.449106   
9996  3.392299  1.274827  2.954947  6.894759  4.349512 -1.663661 -0.952437   
9997  2.364034  2.842030  8.776391  1.008906  4.299976 -1.380719 -0.943884   
9998  9.631511  3.994398  2.757071  7.821347  2.514755 -0.966330 -0.649915   
9999  6.530527  6.781790  4.349695  8.673138  3.492807 -1.390285 -1.532193   

            p4        g1        g2        g3        g4      stab     stabf  
0    -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1    -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2    -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3    -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4    -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable  
...        ...       ...       ...       ...       ...       ...       ...  
9995 -1.236256  0.601709  0.779642  0.813512  0.608385  0.023892  unstable  
9996 -1.733414  0.502079  0.567242  0.285880  0.366120 -0.025803    stable  
9997 -1.975373  0.487838  0.986505  0.149286  0.145984 -0.031810    stable  
9998 -0.898510  0.365246  0.587558  0.889118  0.818391  0.037789  unstable  
9999 -0.570329  0.073056  0.505441  0.378761  0.942631  0.045263  unstable  

[10000 rows x 14 columns]

In [68]:
#View the Shape
df.shape

(10000, 14)

In [69]:
#View the statistical summary
df.describe()

tau1          tau2          tau3          tau4            p1  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean       5.250000      5.250001      5.250004      5.249997      3.750000   
std        2.742548      2.742549      2.742549      2.742556      0.752160   
min        0.500793      0.500141      0.500788      0.500473      1.582590   
25%        2.874892      2.875140      2.875522      2.874950      3.218300   
50%        5.250004      5.249981      5.249979      5.249734      3.751025   
75%        7.624690      7.624893      7.624948      7.624838      4.282420   
max        9.999469      9.999837      9.999450      9.999443      5.864418   

                 p2            p3            p4            g1            g2  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean      -1.250000     -1.250000     -1.250000      0.525000      0.525000   
std        0.433035      0.433035      0.433035      0.274256      0.274255   
min       -1.999891     -1.999945     -1.999926      0.050009      0.050053   
25%       -1.624901     -1.625025     -1.624960      0.287521      0.287552   
50%       -1.249966     -1.249974     -1.250007      0.525009      0.525003   
75%       -0.874977     -0.875043     -0.875065      0.762435      0.762490   
max       -0.500108     -0.500072     -0.500025      0.999937      0.999944   

                 g3            g4          stab  
count  10000.000000  10000.000000  10000.000000  
mean       0.525000      0.525000      0.015731  
std        0.274255      0.274255      0.036919  
min        0.050054      0.050028     -0.080760  
25%        0.287514      0.287494     -0.015557  
50%        0.525015      0.525002      0.017142  
75%        0.762440      0.762433      0.044878  
max        0.999982      0.999930      0.109403

In [70]:
#View the datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    10000 non-null  float64
 1   tau2    10000 non-null  float64
 2   tau3    10000 non-null  float64
 3   tau4    10000 non-null  float64
 4   p1      10000 non-null  float64
 5   p2      10000 non-null  float64
 6   p3      10000 non-null  float64
 7   p4      10000 non-null  float64
 8   g1      10000 non-null  float64
 9   g2      10000 non-null  float64
 10  g3      10000 non-null  float64
 11  g4      10000 non-null  float64
 12  stab    10000 non-null  float64
 13  stabf   10000 non-null  object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [71]:
#View the all types of values contained under column stabf
print(df['stabf'].value_counts())

unstable    6380
stable      3620
Name: stabf, dtype: int64


In [26]:
#Install XGBOOST library using the commented command below
#!pip install xgboost

In [7]:
#Install LIGTHGBM library using the commented command below
#!pip install lightgbm

In [72]:
#Check for missing values
df.isna().sum()

tau1     0
tau2     0
tau3     0
tau4     0
p1       0
p2       0
p3       0
p4       0
g1       0
g2       0
g3       0
g4       0
stab     0
stabf    0
dtype: int64

In [73]:
#Drop stab column
new_df = df.drop(columns=['stab'])

In [74]:
#View changes made to the table
new_df

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
...        ...       ...       ...       ...       ...       ...       ...   
9995  2.930406  9.487627  2.376523  6.187797  3.343416 -0.658054 -1.449106   
9996  3.392299  1.274827  2.954947  6.894759  4.349512 -1.663661 -0.952437   
9997  2.364034  2.842030  8.776391  1.008906  4.299976 -1.380719 -0.943884   
9998  9.631511  3.994398  2.757071  7.821347  2.514755 -0.966330 -0.649915   
9999  6.530527  6.781790  4.349695  8.673138  3.492807 -1.390285 -1.532193   

            p4        g1        g2        g3        g4     stabf  
0    -1.723086  0.650456  0.859578  0.887445  0.958034  unstable  
1    -1.255012  0.413441  0.862414  0.562139  0.781760    stable  
2    -0.920492  0.163041  0.766689  0.839444  0.109853  unstable  
3    -0.997374  0.446209  0.976744  0.929381  0.362718  unstable  
4    -0.554305  0.797110  0.455450  0.656947  0.820923  unstable  
...        ...       ...       ...       ...       ...       ...  
9995 -1.236256  0.601709  0.779642  0.813512  0.608385  unstable  
9996 -1.733414  0.502079  0.567242  0.285880  0.366120    stable  
9997 -1.975373  0.487838  0.986505  0.149286  0.145984    stable  
9998 -0.898510  0.365246  0.587558  0.889118  0.818391  unstable  
9999 -0.570329  0.073056  0.505441  0.378761  0.942631  unstable  

[10000 rows x 13 columns]

In [75]:
#Import numpy library
import numpy as np
#Modify target variable to have machine-readable binary values
new_df['stabf'] = np.where(new_df['stabf']=='stable', 1, 0)

In [85]:
#Declare variables X and y
X = new_df.drop(columns='stabf')
y = new_df['stabf']

In [86]:
#Split data into 80-20 train-test split with random state of 1
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)
y_train.value_counts()

0    5104
1    2896
Name: stabf, dtype: int64

In [89]:
#Use standard scaler to transform the train set and the test set
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train, y_train)
X_test = scaler.transform(X_test)

C:\Users\hp\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [90]:
#Import the relevant libraries needed for Classification Performance Measurement
from sklearn.metrics import recall_score, accuracy_score, precision_score, f1_score, confusion_matrix, classification_report

In [91]:
#Use scikit learn to train a random forest and extra trees classifier
#Random Forest
#Import Random Forest Classifier
from sklearn.ensemble import RandomForestClassifier

#Instantiate and fit the Random Forest Classifier
clf = RandomForestClassifier(n_estimators=100, random_state=1)
clf.fit(X_train, y_train)

#Make preditions for the test set
y_pred = clf.predict(X_test)

#View Accuracy score
accuracy_score(y_test, y_pred)

C:\Users\hp\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


0.362

In [92]:
#Extra Tree Classifier
#Import Random Forest Classifier
from sklearn.ensemble import ExtraTreesClassifier

#Instatiate and fit the Extra Tree Classifier
etc = ExtraTreesClassifier(n_estimators=100, random_state=1)
etc.fit(X_train, y_train)

#Make predictions for the test set
y_pred = etc.predict(X_test)

#View Accuracy
accuracy_score(y_test, y_pred)

C:\Users\hp\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but ExtraTreesClassifier was fitted with feature names
  warnings.warn(


0.3635

In [93]:
#Create a function to train a given model, generate predictions and return accuracy score
def fit_evaluate_model(model, X_train, y_train, X_test, y_test):
    model.fit(X_train, y_train)
    y_predicted = model.predict(X_test)
    return accuracy_score(y_test, y_predicted)

#### Question 2

In [94]:
#Use xgboost and lightgbm to train an extreme boosting model and a light gradient boosting model
#XGBoost

from xgboost import XGBClassifier
xgb_classifier = XGBClassifier(random_state=1)
xgb_accuracy = fit_evaluate_model(xgb_classifier, X_train, y_train, X_test, y_test)
xgb_accuracy

C:\Users\hp\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[18:47:02] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


0.362

In [95]:
#LightGBM Classifier
from lightgbm import LGBMClassifier
lgbm_classifier = LGBMClassifier(random_state=1)
lgbm_accuracy = fit_evaluate_model(lgbm_classifier, X_train, y_train, X_test, y_test)
lgbm_accuracy

0.362

In [82]:
#Obtain the values for confusion matrix, precision, recall, f1 and accuracy
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

[[   0 1288]
 [   0  712]]
              precision    recall  f1-score   support

           0       0.00      0.00      0.00      1288
           1       0.36      1.00      0.53       712

    accuracy                           0.36      2000
   macro avg       0.18      0.50      0.26      2000
weighted avg       0.13      0.36      0.19      2000

0.356


C:\Users\hp\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### ANSWERS TO CODING-RELATED QUESTIONS

### Question 1
Unanticipated. No code was used to attempt question

### Question 2
XGBoost Classifier Answer Obtained: 0.3620

### Question 4
Question was not seen properly

### Question 7
LGBM Classifier. Kernel did not stop loading. But 0.3620 was obtained initially

### Question 12
Random Forest Classifier Answer Obtained: 0.3620

### End of Coding-Related Questions.

### Thank you for the learning opportunity.

### Feedback: Though, I worked very hard and studied the class notes, I felt like I hadn't learnt anything in this Stage after the Quiz. Kindly recommend relevant materials for first-time coders on the Slcak Channel.